### DISCUSSION TITLE
Is Pycoral outdated?

### DISCUSSION THREAD

**Discussion by gnaegit** [geezacoleman/OpenWeedLocator#141] [Q&A]
I followed the instructions to install the OWL software on a Raspberry Pi 4 running Raspberry Pi OS. By default, Python 3.11 is installed. I then tried to install Pycoral to get the GoG algorithm working, but Pycoral only supports Python versions 3.6-3.9. When I checked Pycoral, I noticed that the latest release was in July 2021. This makes me concerned about the likelihood of future support for newer Python versions.

What do you think about this? Is Pycoral (and Google Coral) becoming outdated? Is it still a good choice, or is there a lack of development because there are better alternatives and Google Coral can no longer compete?

https://github.com/google-coral/pycoral/issues/137

**Comment by geezacoleman** [comment]
Thanks for trying it out - when I added the support for PyCoral I felt reasonably confident Google would continue to support it. In the intervening period however and with the launch of the Pi 5, I've become a lot less optimistic about it to the point where I should probably remove the GoG PyCoral section for the reasons you've identified.

It is possible to set it up and use it still with something like the [Hat AI from Pineboards](https://pineboards.io/products/hat-mpcie-coral-edge-tpu-bundle-for-raspberry-pi-5) for improved performance, for the Pi 5 at least running with the PCIe port. The best tutorial I've found is this one from [Jung](https://www.youtube.com/watch?v=HMeDLSflCA0) on YouTube, though it was made before that latest fork you linked to in issue 137.

In any case, for the Pi 4, the PyCoral will help considerably and the lack of PCIe port limits your options to USB-based devices. I've had it working on my Pi 4 previously and it worked fine, but that was before all the new OS/Python 3.11 compatibility issues. On the Pi 5, it's possible to run low resolution, small object detection models at ~ok FPS without any additional hardware. Plus there are new non-Google, Raspberry Pi supported options like the [Hailo AI kit](https://pineboards.io/products/ai-bundle-hailo-8l) and I'm sure others will no doubt follow.

So if you're keen on deploying GoG models I'd recommend upgrading to a Pi 5 - it will give you a lot more flexibility. Fully supported GoG implementation for the OWL is in development and planned for release in the next 6 months, but will only be for Pi 5 and not with the Google Coral. 

I will update the GoG section to reflect this now, so people aren't led astray, and I hope it isn't too frustrating for any time you've already invested in using the Coral with OWL. This was meant to be updated a lot earlier than now (a good 6 month delay) but life gets in the way sometimes of the best intentions!

**Comment by ElectroS3B** [comment]
Hello, I wanted to know where is the GoG implementation for use without Google Corail on rpi 5? Is it still relevant?
I have a rpi5 as well as an ai rpi camera. In GoB it seems to work well, but I have only tested it in the office. I would like to use it to do GoG but I don't know how to do it.

**Reply by geezacoleman** [reply]
I'm still working on the ultralytics implementation for the Pi5, but it works for me currently and is over on [ultralytics-gog](https://github.com/geezacoleman/OpenWeedLocator/tree/ultralytics-gog) branch.

It will work without any additional AI accelerator and you can get up to 12FPS which should be fine for most applications 10kmh and under. This approach is probably the easiest for GoG implementation with custom models. Training with YOLO is much easier, you'll just need to export in NCNN format for faster runtimes.

I'll be working on this branch in the coming weeks to test and add documentation so it's more clear how to use it. Hopefully this helps, let me know how you go with the AI camera as well, that requires a different approach.

**Reply by ElectroS3B** [reply]
Once again thank you very much for your answer and the time you dedicate to me. I will try to use the GoG branch. Regarding my work with the ai rpi camera, I encounter a difficulty in converting a trained Yolov8n.pt model to imx format which is the format accepted by the ia imx500 module of the camera. I have no programming skills and I use tools that I discover (python, google colab ...), my difficulty comes from there, but I do not despair, I have already advanced a little by understanding and training a yolov8 model with photos found on the net and on weedai. I persevere, I have the impression that this little camera has a lot to show us ;-)

**Reply by geezacoleman** [reply]
No worries, happy to help where I can and it's great to see that weedai is helpful.

The AI camera would be quite a neat solution. Those model conversion issues are why the Hailo modules are also a bit of a hassle
